test/boost/big_decimal_test.cc

```diff
@@ -203,3 +203,118 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    };
+    auto test_cmp = [&] (const big_decimal& a, const big_decimal& b, std::strong_ordering expected_result, std::source_location sl = std::source_location::current()) {
+        fmt::print("checking {} {} {} @ {}:{}\n", fmt_decimal(a), fmt_ordering(expected_result), fmt_decimal(b), sl.file_name(), sl.line());
+        auto res = a <=> b;
+        if (res != expected_result) {
+            BOOST_FAIL(fmt::format("{} <=> {}: expected result: {}, but got {}", fmt_decimal(a), fmt_decimal(b), fmt_ordering(expected_result), fmt_ordering(res)));
+        }
+    };
+
+    // Test compare with same scale, should hit early return for same scale.
+    test_cmp(big_decimal("1"), big_decimal("-1"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.3345E3"), big_decimal("-1.3344E3"), std::strong_ordering::greater);
+    test_cmp(big_decimal("18E2"), big_decimal("2E2"), std::strong_ordering::greater);
+
+    // Test compare with different signs, should hit early return for different signs.
+    test_cmp(big_decimal("-1.3345E3"), big_decimal("1"), std::strong_ordering::less);
+    test_cmp(big_decimal("1.3345E3"), big_decimal("-1.3344E5"), std::strong_ordering::greater);
+
+    // Test compare against 0, should hit early return for different zero.
+    test_cmp(big_decimal("0"), big_decimal("0"), std::strong_ordering::equal);
+    test_cmp(big_decimal("0E6"), big_decimal("0E1"), std::strong_ordering::equal);
+    test_cmp(big_decimal("1"), big_decimal("0"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1"), big_decimal("0E7"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.32432345E7"), big_decimal("0"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1"), big_decimal("0"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1"), big_decimal("0E-4"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.3345E3"), big_decimal("0"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("1"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("187687E3"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("-187687E3"), std::strong_ordering::greater);
+
+    // Test compare scale-compare algorithm.
+    test_cmp(big_decimal("1.1221310000000023424234E9"), big_decimal("1.267687687968768686"), std::strong_ordering::greater);
+    test_cmp(big_decimal("1.267687687968768686"), big_decimal("1.1221310000000023424234E9"), std::strong_ordering::less);
+    test_cmp(big_decimal("134252342E9"), big_decimal("100023424234E19"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.1221310000000023424234E9"), big_decimal("-1.267687687968768686"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.267687687968768686"), big_decimal("-1.1221310000000023424234E9"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-134252342E9"), big_decimal("-100023424234E19"), std::strong_ordering::greater);
```

Comment:
These test don't stress the area where the numbers are close, but the scales are very far.

Examples:

1e20 vs "9" * 20      (python syntax for repeating "9")
1e20 vs "1" "0"*19 "1"

Reply:
I see below you have tests that to test it.